From 93e082e0b6bdfdbbbb3075153005b48b17516f64 Mon Sep 17 00:00:00 2001
From: "xinqi.wang" <XinQi.Wang@cixtech.com>
Date: Fri, 17 Jan 2025 20:36:18 +0800
Subject: [PATCH] DPTSW-11888: v4l2: move buffer state fetch into pool lock

When seek in clapper playback, in streamoff operation, a separate output
buffer will be queued and may take one of the buffer index in later
buffers which will also be queued by VPU stream-off IOCTL;
Move buffer state fetch into pool lock in streamoff and qbuf to
guarantee:
clapper frame buffer queue, which is from gst_buffer_pool_release_buffer
will not mix into VPU frame buffers queue after streamoff IOCTL

Signed-off-by: Xinqi.Wang <xinqi.wang@cixtech.com>
Change-Id: I6d95c9be40710fccc58b91e8b065c1fd56fe0452

diff --git a/sys/v4l2/gstv4l2bufferpool.c b/sys/v4l2/gstv4l2bufferpool.c
index e7a2946994..92b8966620 100644
--- a/sys/v4l2/gstv4l2bufferpool.c
+++ b/sys/v4l2/gstv4l2bufferpool.c
@@ -798,6 +798,7 @@ gst_v4l2_buffer_pool_streamoff (GstV4l2BufferPool * pool)
   GstBufferPoolClass *pclass = GST_BUFFER_POOL_CLASS (parent_class);
   GstV4l2Object *obj = pool->obj;
   gint i;
+  gint buffer_state[VIDEO_MAX_FRAME];
 
   if (!pool->streaming)
     return;
@@ -811,6 +812,10 @@ gst_v4l2_buffer_pool_streamoff (GstV4l2BufferPool * pool)
       if (obj->ioctl (pool->video_fd, VIDIOC_STREAMOFF, &obj->type) < 0)
         GST_WARNING_OBJECT (pool, "STREAMOFF failed with errno %d (%s)",
             errno, g_strerror (errno));
+      for (i = 0; i < VIDEO_MAX_FRAME; i++) {
+        buffer_state[i] =
+          g_atomic_int_and (&pool->buffer_state[i], ~BUFFER_STATE_QUEUED);
+      }
 
       pool->streaming = FALSE;
 
@@ -824,15 +829,13 @@ gst_v4l2_buffer_pool_streamoff (GstV4l2BufferPool * pool)
   }
 
   for (i = 0; i < VIDEO_MAX_FRAME; i++) {
-    gint old_buffer_state =
-        g_atomic_int_and (&pool->buffer_state[i], ~BUFFER_STATE_QUEUED);
-    if ((old_buffer_state & BUFFER_STATE_QUEUED) && pool->buffers[i]) {
+    if ((buffer_state[i] & BUFFER_STATE_QUEUED) && pool->buffers[i]) {
       GstBuffer *buffer = pool->buffers[i];
       GstBufferPool *bpool = GST_BUFFER_POOL (pool);
 
       pool->buffers[i] = NULL;
 
-      if (!(old_buffer_state & BUFFER_STATE_OUTSTANDING)) {
+      if (!(buffer_state[i] & BUFFER_STATE_OUTSTANDING)) {
         if (V4L2_TYPE_IS_OUTPUT (pool->obj->type))
           gst_v4l2_buffer_pool_complete_release_buffer (bpool, buffer, FALSE);
 
@@ -1238,14 +1241,6 @@ gst_v4l2_buffer_pool_qbuf (GstV4l2BufferPool * pool, GstBuffer * buf,
 
   index = group->buffer.index;
 
-  old_buffer_state =
-      g_atomic_int_or (&pool->buffer_state[index], BUFFER_STATE_QUEUED);
-  if (old_buffer_state & BUFFER_STATE_QUEUED)
-    goto already_queued;
-
-  GST_LOG_OBJECT (pool, "queuing buffer %i, previous-state = %i", index,
-      old_buffer_state);
-
   if (V4L2_TYPE_IS_OUTPUT (obj->type)) {
     enum v4l2_field field;
 
@@ -1271,6 +1266,14 @@ gst_v4l2_buffer_pool_qbuf (GstV4l2BufferPool * pool, GstBuffer * buf,
     }
   }
 
+  old_buffer_state =
+      g_atomic_int_or (&pool->buffer_state[index], BUFFER_STATE_QUEUED);
+  if (old_buffer_state & BUFFER_STATE_QUEUED)
+    goto already_queued;
+
+  GST_LOG_OBJECT (pool, "queuing buffer %i, previous-state = %i", index,
+      old_buffer_state);
+
   /* If the pool was orphaned, don't try to queue any returned buffers.
    * This is done with the objet lock in order to synchronize with
    * orphaning. */
